qualtran/bloqs/bookkeeping/auto_partition.py

```diff
@@ -0,0 +1,116 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from itertools import chain
+from typing import Dict, Sequence, Set, Tuple
+
+from attrs import evolve, field, frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QAny,
+    Register,
+    Side,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.bookkeeping.partition import Partition
+from qualtran.cirq_interop.t_complexity_protocol import TComplexity
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class AutoPartition(Bloq):
+    """Wrap a bloq with `Partition` to fit an alternative set of input and output registers
+       such that splits / joins can be avoided in diagrams.
+
+    Args:
+        bloq: The bloq to wrap.
+        partition: A sequence of pairs specifying each register that the wrapped bloq should accept
+            and the registers from `bloq.signature.lefts()` that concatenate to form it.
+        partition_output: If True, the output registers will also follow `partition`.
+            Otherwise, the output registers will follow `bloq.signature.rights()`.
```

Comment:
what's the motivation for this flag? I would have assumed the `False` setting would be most useful so you could fully wrap a bloq's factorized registers

Reply:
This flag is not important if we're only wrapping Bloqs with the same left and right registers. It could be useful in cases where we only want to insert a `Partition` on the left side but keep the original signature of the bloq on the right side. In that case, `partition_output` should be False. The default case, where both sides are wrapped, is True.

Reply:
Oh, I read the documentation and variable name backwards. I think of the "partitioned" form of registers as being the many, split-up registers and the unpartitioned form being the single register. So when I read `partition_output = True`, I assume the output will be left in the partitioned state. 

Maybe a completely different flag toggle like `left_only = False` or something like that. It's probably also useful to capture what you wrote in the above comment ^ in the docstring somewhere, i.e. that (full) wrapping only works for consistent left&right registers